[#12] Adding scheduled shutdown

DT3264 opened the pull request:
[#12] Adding scheduled shutdown





This commit adds support for a new global directive "ShutdownTime", like ContestTimes or ConfigExpirationTime, that schedules a shutdown at the time defined.

equetzal commented on base-system/usrroot/usr/lib/hsync/libhapply.so:
```diff
@@ -47,6 +47,7 @@ apply_global_directives() {
 	set_timezone
 	set_keyboards
 	set_current_keyboard
+	set_scheduled_shutdown
```
Let's condition the shutdown to the clock being NTP sync. Otherwise this could lead computers to shutdown in an unwanted moment if the hardware clock is out of sync. 
This is important to also be updated on the documentation.

DT3264 replied:
Updating that

equetzal commented on base-system/usrroot/usr/lib/hsync/libhapply.so:
```diff
@@ -404,3 +405,22 @@ set_wallpaper() {
+	fi
+	MINUTES_TO_SHUTDOWN=$((SECONDS_TO_SHUTDOWN/60))
+	log "Scheduling shutdown in $MINUTES_TO_SHUTDOWN minute(s)"
+	shutdown +"$MINUTES_TO_SHUTDOWN"
```
Why not to schedule this with the `at` command?

equetzal replied:
If we end using this maybe the only problem could be erasing scheduled applies.

DT3264 replied:
> Why not to schedule this with the `at` command?

I think is less prone to unexpected shutdowns because based on [this](https://unix.stackexchange.com/a/465323) someone says

> if you use at, the job will be stored and if the system is shutdown by some other means before 18:00, upon booting again the job will still be waiting to be run; if the time is already passed, the shutdown will be performed immediately which could be quite unexpected.

DT3264 replied:
> If we end using this maybe the only problem could be erasing scheduled applies.

You mean the at, or the shutdown command? :thinking:

equetzal replied:
> > Why not to schedule this with the `at` command?
> 
> I think is less prone to unexpected shutdowns because based on [this](https://unix.stackexchange.com/a/465323) someone says
> 
> > if you use at, the job will be stored and if the system is shutdown by some other means before 18:00, upon booting again the job will still be waiting to be run; if the time is already passed, the shutdown will be performed immediately which could be quite unexpected.

Interesting, then what happens if the system changes the clock. Let's suppose that system have 10am and shutdown is scheduled to shutdown at 12pm, then the system clock changes to 11am. The system will shutdown at 12pm or at 1pm?

DT3264 replied:
After some testings, seems like in the example you mention, the system will still shutdown at 12pm as scheduled without being affected by the clock changing.

equetzal commented on base-system/usrroot/usr/lib/hsync/libhapply.so:
```diff
@@ -404,3 +405,22 @@ set_wallpaper() {
 	return 1 # error
 }
+
+set_scheduled_shutdown() {
```
Let's try to make variables `local` instead of global. 
It could be something like
`local SHUTDOWN_TIME SHUTDOWN_TIME_EPOCH CURRENT_TIME_EPOCH SECONDS_TO_SHUTDOWN MINUTES_TO_SHUTDOWN`

DT3264 replied:
Done

equetzal commented on docs/usage/directives/configurations/shutdown-time.md:
```diff
@@ -0,0 +1,12 @@
+
+`ShutdownTime=[ ISO8601_TIME | none ]`
+
+Allows to set a specific time in the global section when the instances should shutdown.
```
Let's specify on here that system will only shutdown if the systemclock is properly in sync with an NTP server.

DT3264 replied:
Okay, added.

equetzal commented on docs/usage/directives/configurations/shutdown-time.md:
```diff
@@ -0,0 +1,12 @@
+
+`ShutdownTime=[ ISO8601_TIME | none ]`
+
+Allows to set a specific time in the global section when the instances should shutdown (only if the system clock is properly sync with an NTP server).
```
I think we can ask chatGPT to rewrite this in a more professional way and also we can set more details of what to expect on date adjustments, etc.

DT3264 replied:
Done

equetzal reviewed (approved):
Have not tested, please rock-test it before landing. 
Address de documentation comment. LGTM in overall!

equetzal commented on base-system/usrroot/usr/lib/hsync/libhapply.so:
```diff
@@ -404,3 +406,23 @@ set_wallpaper() {
+		log "Current shutdown is for a time that already happened, skiping scheduled shutdown."
+		exit 0
+	fi
+	MINUTES_TO_SHUTDOWN=$((SECONDS_TO_SHUTDOWN/60))
```
Let's do a rounding here. For a minute, if less than 30s then 0min (floor), if more than 30s then 1min (ceil). E.g. 1:45 then the system will shutdown in 2min. 

This is because I noticed the following: Scheduled a system mode change on 00:05 min, and shutdown at 00:06 min. Then on the 00:05 sync, the system directly shutdown without firstly changing the mode, this was because at the time of the check, the seconds to shutdown were 54s, which leaded to shutdown now.

DT3264 replied:
Okay, added a rounding.

equetzal commented on base-system/usrroot/usr/lib/hsync/libhapply.so:
```diff
@@ -404,3 +406,23 @@ set_wallpaper() {
+	fi
+	MINUTES_TO_SHUTDOWN=$((SECONDS_TO_SHUTDOWN/60))
+	log "Scheduling shutdown in $MINUTES_TO_SHUTDOWN minute(s)"
+	shutdown +"$MINUTES_TO_SHUTDOWN"
```
Also, let's consider the option to use [*systemd-shutdownd*](https://www.freedesktop.org/software/systemd/man/systemd-poweroff.service.html). It might have some advantages over *shutdown*.

DT3264 replied:
On this side fater some searching seems like `shutdown` is a link to `systemd-shutdown`, so they do the same thing. Unless there's something I'm not seeing, feel free to mention if that's the case

equetzal reviewed (changes requested):
Overall good, the system does shutdown correctly and does the work. 
I've found some problems that are needed to be corrected before merging, so let's tackle them. 
Good work on to this! :D

equetzal commented:
Hey @DT3264, I was notified for review on this, but I don't see any upcoming change after my last review. Was this request right?

DT3264 commented:
@equetzal There's no need to re-review this one yet, dunno why that review was requested

DT3264 commented on docs/usage/directives/configurations/shutdown-time.md:
```diff
@@ -0,0 +1,12 @@
+# Shutdown time
+
+`ShutdownTime=[ ISO8601_TIME | none ]`
```
Done

DT3264 commented on base-system/usrroot/usr/lib/hsync/libhapply.so:
```diff
@@ -404,3 +406,23 @@ set_wallpaper() {
+	SHUTDOWN_TIME="$(get_directives_var Global ShutdownTime)"
+	log "Cancelling scheduled shutdown (if any)"
+	shutdown -c
+	if [ "$SHUTDOWN_TIME" = "none" ]; then
```
Although gonna validate that on the directives validator, I added a failsafe so if SHUTDOWN_TIME is empty, it will fallback to none

DT3264 commented on base-system/usrroot/usr/lib/hsync/libhapply.so:
```diff
@@ -404,3 +406,23 @@ set_wallpaper() {
+	SECONDS_TO_SHUTDOWN=$((SHUTDOWN_TIME_EPOCH-CURRENT_TIME_EPOCH))
+	if [ "$SECONDS_TO_SHUTDOWN" -lt 0 ]; then
+		log "Current shutdown is for a time that already happened, skiping scheduled shutdown."
+		exit 0
```
Done

DT3264 commented:
@equetzal I asked for review intentionally, I amended  the requested changes

equetzal commented on docs/usage/directives/configurations/shutdown-time.md:
I'll test this ASAP but in the meanwhile please rework this doc similar as you did with ipman. Try to be more detailed about what this is doing, talk about the roundings of the time, explain what happens on edge cases for someone to be clear about what happens in those cases. 
Remember that the whole point of documentation if to find the specifics on how something works without requiring someone to got deep on the code. An example of this was the discovery we had about the `date %s ''` which had that 'special' case of empty string, and we catch that on the docs not the code.

DT3264 committed:
[#12] Adding scheduled shutdown





This commit adds support for a new global directive "ShutdownTime", like ContestTimes or ConfigExpirationTime, that schedules a shutdown at the time defined.

equetzal reviewed (changes requested):
I've found several issues so I'll be needing to hold this feature because it's not ready for a release. The bug of not being able to change mode and how hard is to recover from it makes this severe, so we might be needing to fully change the approach we're taking for this feature.
Also, you need to polish your testing skills, you're not being able to find multiple bugs in several PRs that you craft. I know this environment is not like regular coding, this is why you need to become creative when it's about testing and really try your feature in all kind of circunstancies. 

Great work on this, but we need to keep working on this one.

DT3264 commented:
I think I'll rework this PR later on after the OMI thing has happened, in the meantime I'll leave a thought I had about this feature

Maybe the shutdown thing could be simpler, just a service asking each minute 
"Is the current datetime equal or greater than the shutdown datetime provided?" 
And to avoid conflicts with the mode change, maybe if there's a mode change nearby skip the shutdown until the mode change had taken place, 

wdyt?

equetzal commented:
> I think I'll rework this PR later on after the OMI thing has happened, in the meantime I'll leave a thought I had about this feature
> 
> Maybe the shutdown thing could be simpler, just a service asking each minute "Is the current datetime equal or greater than the shutdown datetime provided?" And to avoid conflicts with the mode change, maybe if there's a mode change nearby skip the shutdown until the mode change had taken place,
> 
> wdyt?

I like the approach of instead of scheduling the shutdown with `shutdown`, do it like "schedule a script execution" which when executed actually powers off the machine. This way we can implement this in numerous ways, we could use `at`, `cron`, `systemd-run`, etc. And the shutdown op would always be like 'shutdown now'. Inside, the op could be a systemd-service that conflicts with both `hsync.service` and `happly.service` so that the execution waits until the directives are applied to then shutdown the system. 

About the thing with mode change, is not -that- important, the problem is the accuracy of the time, because if I do some op at min 0 and shutdown at min 1 I don't expect the system to shutdown before, but this can be solved by the systemd service I told you.
